src/hotspot/share/classfile/stringTable.hpp

```diff
@@ -104,14 +105,43 @@ class StringTable : public CHeapObj<mtSymbol>{
   }
 
   // Sharing
+#if INCLUDE_CDS_JAVA_HEAP
+  static inline oop read_string_from_compact_hashtable(address base_address, u4 index);
+
+private:
+  static bool _is_two_dimensional_shared_strings_array;
+  static OopHandle _shared_strings_array;
+  static int _shared_strings_array_root_index;
+
+  // All the shared strings are referenced through _shared_strings_array to keep them alive.
+  // Each shared string is stored as a 32-bit index in ::_shared_table. The index
+  // is interpreted in two ways:
+  //
+  // [1] _is_two_dimensional_shared_strings_array = false: _shared_strings_array is an Object[].
+  //     Each shared string is stored as _shared_strings_array[index]
+  //
+  // [2] _is_two_dimensional_shared_strings_array = true: _shared_strings_array is an Object[][]
+  //     This happens when there are too many elements in the shared table. We store them
+  //     using two levels of objArrays, such that none of the arrays are too big for
+  //     ArchiveHeapWriter::is_too_large_to_archive(). In this case, the index is splited into two
+  //     parts. Each shared string is stored as _shared_strings_array[primary_index][secondary_index]:
+  //
+  //           [bits 31 .. 14][ bits 13 .. 0  ]
+  //            primary_index  secondary_index
+  const static int _secondary_array_index_bits = 14;
```

Comment:
I wanted `_secondary_array_index_bits` to be a compile-time constant so that the related code can be optimized by the C++ compiler. However, to calculate it at compile time would require the use of `constexpr`, but `objArrayOopDesc::object_size(int)` can't easily be made `constexpr` because it eventually reads the runtime constant `heapOopSize`.

Instead, I added a function `StringTable::verify_secondary_array_index_bits()` to check that its value isn't too big.

Note that I used 14 as that will be the eventually value when we support Shenandoah. Also, using a large value means the SharedStringsStress needs a larger data set, so I stick with 14.

This will still support up to 28M strings so it's should be OK.

Reply:
@iklam  thanks for adding those comments in `StringTable::verify_secondary_array_index_bits`.